### ISSUE TITLE
Peer Review

### ISSUE THREAD

**Issue by alexdifre** [Rishi005/Algortihms-and-AI-Lab#1]
# Code Review, 12/01/2025 , 13:45:46
## Positive Aspects

1. **Good Testing**: Tests like `test_numerical_gradient_check()` and `test_overfits_minibatch()`  demonstrate understanding of neural network validation.

2. **Clear Documentation**: The Big-O analysis is well-explained. 

3. **Code Readability**: Variable names are descriptive, and the overall structure follows the mathematical formulation closely, making it easy to follow.

### the  critical Issue

1. **Hardcoded File Path in `mnist_loader.py`**:
```python
   f = gzip.open(r"/home/rishika/Algorithms_and_AI/Algortihms-and-AI-Lab/NN_project/data/mnist.pkl.gz", 'rb')
```
   This absolute path will break on any other system.  In your code, I would do this :
```
   import os
   data_path = os.path.join(os.path.dirname(__file__), 'data', 'mnist.pkl.gz')
   f = gzip.open(data_path, 'rb')
```

### Significant Improvements

1. **Weight Initialization Could Be Better**: The current Gaussian initialization works but can be improved. Consider Xavier/He initialization:
```python
   self.weights = [np.random.randn(y, x) / np.sqrt(x) for x, y in zip(sizes[:-1], sizes[1:])]
```
   This helps prevent vanishing/exploding gradients, especially with deeper networks.

2. **Learning Rate is Fixed**: Consider implementing learning rate decay for more stable and better convergence
3.  **Different Activation Functions**: Consider implementing ReLU, which often trains faster than sigmoid.

4.  **Regularization**: Add L2 regularization, for example, to prevent overfitting: but it' s totally optional.

5. **Cross-Entropy Loss**: For classification, cross-entropy often works better than MSE.

### Minor Improvements

1. **Type Hints**: Adding type hints could improve code clarity. I personally like them a lots

2. **Cost Function Should Be a Method**: The MSE cost calculation is scattered. 

3. If I understood correctly, the values like `3.0` for learning rate and `30` for hidden layer size appear in multiple places. Consider using defined constants 

Consider separating the data loading logic from the model architecture. A `data_loader.py` would be nice
